### PR TITLE
Add per-project walkthrough history (#9)

### LINKED ISSUE #9: [Feature]: Add per-project walkthrough history and replay
TL;DR: GitHub issue body for per-project walkthrough history.
It records the confirmed phase-1 scope decisions.
The issue uses the existing `enhancement` GitHub label because the repo has no `feature` or size labels.
The implementation plan points to current walkthrough-plugin modules.
Next step: open the issue, record the URL, then implement in a separate worktree.

**Labels:** feature, large

## Status Quo

Walkthrough Plugin currently exposes one MCP tool, `show_walkthrough_items`, implemented in `src/main/kotlin/com/forketyfork/walkthrough/ShowWalkthroughItemsToolset.kt`. The tool accepts a JSON array of walkthrough items and then calls `showWalkthroughItems(project, editor, itemList)` in `WalkthroughOrchestrator.kt`.

The walkthrough is displayed immediately through the existing Compose/Jewel popup flow. Navigation can jump to an optional project-relative file and 1-based line, and the popup supports previous/next navigation while that live session is open.

There is no per-project persistence today. Once the popup is closed, the walkthrough data is discarded, there is no IDE action for reopening previous walkthroughs, and there is no top-level description field that can be shown in a history list.

## Objectives

Persist each walkthrough created for a project so it can be replayed later from the IDE. Users should be able to bind an IDE shortcut to a history action, open a small dropdown for the current project, pick a prior walkthrough by date and description, and replay it from the first step.

Old walkthroughs may point at files or lines that no longer exist. Replay should still be useful in those cases: the plugin should show what it can and degrade gracefully instead of throwing.

## User Flow

**Trigger:** An agent calls `show_walkthrough_items` with a description and one or more walkthrough items.

1. The plugin validates the tool call, shows the walkthrough, and saves the full payload under the current project's `.idea/walkthroughs/` directory.
2. Later, the user invokes the keymap-discoverable walkthrough history action.
3. The plugin opens a lightweight popup listing prior walkthroughs for that project, sorted newest first, with rows like `2026-05-09 06:34 - Explain popup placement`.
4. The user selects a row.
5. The plugin loads the stored walkthrough and replays it through the same popup driver, starting from item zero.

**Result:** Previous walkthroughs are available from the current project and can be replayed even after the original popup is closed.

## Scope

**In scope:**

- Store each completed `show_walkthrough_items` payload under the current project's `.idea/walkthroughs/` directory.
- Persist id, creation timestamp, description, and the full list of walkthrough items.
- Add a required top-level `description` parameter to the MCP tool call.
- Add a project-level history service for save/list/load.
- Add a keymap-discoverable IDE action, without a default keystroke, that opens a list popup for prior walkthroughs.
- Replay selected walkthroughs from the first item by reusing the existing popup flow.
- Handle stale file/line anchors on a best-effort basis without propagating errors.
- Add focused tests for record serialization, history listing/loading, corrupt-file tolerance, and stale-anchor resolution.

**Out of scope:**

- The color palette and settings-panel feature.
- Resume support, checkpoints, or starting replay from the middle of a walkthrough.
- User-global history outside the project.
- A tool window or full management UI for deleting, editing, searching, or renaming walkthroughs.
- Perfect recovery when files move or code is rewritten.
- Bumping `pluginVersion`.

### Affected Modules

- `src/main/kotlin/com/forketyfork/walkthrough/ShowWalkthroughItemsToolset.kt`: add the description parameter, validate it, and save successful walkthrough calls.
- `src/main/kotlin/com/forketyfork/walkthrough/WalkthroughItem.kt`: keep the item model as the shared payload shape for live display and stored records, or add a small serializable DTO if needed.
- `src/main/kotlin/com/forketyfork/walkthrough/WalkthroughOrchestrator.kt`: refactor the replay/display entry point so live tool calls and stored records share the same popup flow.
- `src/main/kotlin/com/forketyfork/walkthrough/WalkthroughHistoryService.kt`: new project service that writes, lists, and loads records from `.idea/walkthroughs/`.
- `src/main/kotlin/com/forketyfork/walkthrough/WalkthroughHistoryAction.kt`: new IDE action that opens the history popup and starts replay.
- `src/main/resources/META-INF/plugin.xml`: register the project service and action under a Walkthrough group for keymap discovery.
- `src/test/kotlin/com/forketyfork/walkthrough/`: add tests for storage and anchor degradation.

### Tasks

Each task is independently testable. Ordered by implementation sequence.

1. [ ] Add a persisted `WalkthroughRecord` model with id, created-at timestamp, description, and items — `WalkthroughRecord.kt`
2. [ ] Add `WalkthroughHistoryService` with atomic save, newest-first list, load by id, and corrupt-file skip behavior — `WalkthroughHistoryService.kt`
3. [ ] Extend `show_walkthrough_items` with a top-level description parameter and save records after validation — `ShowWalkthroughItemsToolset.kt`
4. [ ] Extract reusable replay/display entry points so live calls and stored records both start at item zero — `WalkthroughOrchestrator.kt`
5. [ ] Add stale-anchor resolution helpers so missing files, out-of-range lines, and moved files never throw during replay — `WalkthroughAnchorResolver.kt`
6. [ ] Add `WalkthroughHistoryAction` and list popup UX with rows showing date and description — `WalkthroughHistoryAction.kt`
7. [ ] Register the project service and keymap-discoverable action — `src/main/resources/META-INF/plugin.xml`
8. [ ] Write tests: record round trip, list/load sorting, corrupt-file skip, and stale-anchor cases
9. [ ] Update docs/PRD.md (if present)
10. [ ] Update docs/TRACEABILITY.md (if present)
11. [ ] Update docs/ARCHITECTURE.md (if architecture changed and file present)

### New Dependencies

None. Use APIs already available to the plugin, including the existing JSON stack used by the MCP toolset.

## Acceptance Criteria

- [ ] All tasks completed
- [ ] Each successful `show_walkthrough_items` call is persisted under the current project's `.idea/walkthroughs/`
- [ ] Stored records include id, creation timestamp, description, and full walkthrough items
- [ ] The history action is discoverable in Settings | Keymap and has no default shortcut
- [ ] The history popup lists prior walkthroughs for the current project by date and description, newest first
- [ ] Selecting a history row replays that walkthrough from item zero
- [ ] Stale file and line anchors degrade gracefully and do not throw
- [ ] Tests cover happy path and documented edge cases
- [ ] docs/PRD.md updated (if present)
- [ ] docs/TRACEABILITY.md updated (if present)
- [ ] Pre-commit hooks pass

## Confirmed Scope Decisions

- On-disk format: `.idea/walkthroughs/<timestamp>-<slug>.json` per walkthrough, with an optional small `index.json` only if listing performance needs it.
- VCS behavior: no plugin-added `.gitignore` entry for `.idea/walkthroughs/`; the folder is trackable by default.
- Action surface: `WalkthroughHistoryAction` with action id `com.forketyfork.walkthrough.ShowHistory`, under a `Walkthrough` keymap group, with no default shortcut.
- Dropdown UX: lightweight list popup using IntelliJ popup APIs, anchored to IDE focus owner, with rows showing `date - description`.
- MCP schema: single required top-level `description: String` parameter.
- Stale-anchor policy: show the walkthrough anyway; if the file or line cannot resolve, keep replay running and show the step at the current editor/caret fallback position.

### PR DESCRIPTION
**Pull request by forketyfork**
Closes #9.

Adds per-project walkthrough history under `.idea/walkthroughs/`, with one JSON file per walkthrough and an IDE action users can bind in Keymap. The MCP tool now requires a concise `description`, saves shown walkthroughs, and the replay path reuses the existing popup flow.

Stale anchors are best-effort: missing files or invalid lines fall back to the current editor/caret instead of throwing. Corrupt history files are skipped when listing.

Verification:
- `just lint`
- `just test`
- `just build`

### CODE REVIEW

File: src/main/kotlin/com/forketyfork/walkthrough/ResolvedWalkthroughTarget.kt [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    item: WalkthroughItem
+): ResolvedWalkthroughTarget? {
+    val editor = fileEditorManager.selectedTextEditor ?: fallbackEditor ?: return null
+    val popupItem = if (isResolvableWalkthroughLine(item.line, editor.document.lineCount)) {
+        moveCaretToLine(editor, item.line)
+        item
+    } else {
+        item.withFallbackAnchor()
+    }
+    return ResolvedWalkthroughTarget(editor, popupItem)
```
[the remark this answers is not in the document]
Comment (forketyfork) [reply src/main/kotlin/com/forketyfork/walkthrough/ResolvedWalkthroughTarget.kt:42]: Fixed in `cc904a6`. If a stored item points to a file that no longer resolves, replay now clears the stale file and line before falling back, so the popup stays on the current editor/caret.

File: src/main/kotlin/com/forketyfork/walkthrough/ResolvedWalkthroughTarget.kt [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    } else {
+        null
+    }
+    return editor?.let { ResolvedWalkthroughTarget(it, item) }
+}
+
+private fun findWalkthroughFile(project: Project, relativePath: String) =
+    project.basePath
+        ?.let { "$it/$relativePath" }
+        ?.let(LocalFileSystem.getInstance()::findFileByPath)
```
[the remark this answers is not in the document]
Comment (forketyfork) [reply src/main/kotlin/com/forketyfork/walkthrough/ResolvedWalkthroughTarget.kt:64]: Fixed in `cc904a6`. File targets are now resolved with `Path` normalization: absolute paths are rejected, traversal is blocked, and only paths under the project root can be opened.

File: src/main/kotlin/com/forketyfork/walkthrough/WalkthroughHistoryStore.kt [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+        return Files.list(directory).use { paths ->
+            paths
+                .filter(::isRecordFile)
+                .map { path -> readRecordOrNull(path) }
+                .toList()
+                .filterNotNull()
+                .sortedWith(
+                    compareByDescending<WalkthroughRecord> { record -> record.createdAtInstantOrEpoch() }
+                        .thenByDescending { record -> record.id }
+                )
```
[the remark this answers is not in the document]
Comment (forketyfork) [reply src/main/kotlin/com/forketyfork/walkthrough/WalkthroughHistoryStore.kt]: Fixed in `cc904a6`. `list()` now catches directory-listing failures, reports them through the existing history error callback, and returns an empty list instead of bubbling the error into the action.

File: src/main/kotlin/com/forketyfork/walkthrough/WalkthroughHistoryService.kt [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    }
+
+    fun save(description: String, items: List<WalkthroughItem>): WalkthroughRecord? =
+        store?.save(description, items)
+
+    fun list(): List<WalkthroughRecord> =
+        store?.list().orEmpty()
+
+    fun load(id: String): WalkthroughRecord? =
+        store?.load(id)
```
[the remark this answers is not in the document]
Comment (forketyfork) [reply src/main/kotlin/com/forketyfork/walkthrough/WalkthroughHistoryService.kt]: Fixed in `cc904a6`. The service now wraps save/list/load, logs failures, and returns `null` or an empty list so storage problems do not break the action or MCP flow.

File: src/main/kotlin/com/forketyfork/walkthrough/ShowWalkthroughItemsToolset.kt [hunk 5]
```diff
@@ -62,48 +61,14 @@ class ShowWalkthroughItemsToolset : McpToolset {
...
-        ?.let(LocalFileSystem.getInstance()::findFileByPath)
+        val record = WalkthroughHistoryService.getInstance(project)
+            .save(trimmedDescription, itemList)
+            ?: mcpFail("Project path is required to save walkthrough history")
 
-private fun moveCaretIfNeeded(editor: Editor?, line: Int?): Editor? {
-    if (editor != null && line != null) {
-        editor.caretModel.moveToLogicalPosition(LogicalPosition(line - 1, 0))
-        editor.scrollingModel.scrollToCaret(ScrollType.CENTER)
+        return "Walkthrough items shown and saved as ${record.id}"
```
[the remark this answers is not in the document]
Comment (forketyfork) [reply src/main/kotlin/com/forketyfork/walkthrough/ShowWalkthroughItemsToolset.kt]: Fixed in `cc904a6`. Showing a walkthrough remains the primary success path; history persistence is best-effort, and the MCP response now reports success even when saving was skipped or failed.

File: src/main/kotlin/com/forketyfork/walkthrough/ResolvedWalkthroughTarget.kt [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+private fun resolveFallbackTarget(
+    fileEditorManager: FileEditorManager,
+    fallbackEditor: Editor?,
+    item: WalkthroughItem
+): ResolvedWalkthroughTarget? {
+    val editor = fileEditorManager.selectedTextEditor ?: fallbackEditor ?: return null
+    val popupItem = if (isResolvableWalkthroughLine(item.line, editor.document.lineCount)) {
+        moveCaretToLine(editor, item.line)
+        item
```
[the remark this answers is not in the document]
Comment (forketyfork) [reply src/main/kotlin/com/forketyfork/walkthrough/ResolvedWalkthroughTarget.kt:38]: Fixed in `cc904a6`. This is handled by the same fallback change: unresolved file targets now clear file and line before replay falls back to the current editor/caret.
